### PR TITLE
e4s ci: unify when possible; incorporate ecp-data-vis-sdk metapackage [do not merge yet]

### PR DESCRIPTION
**Pull request by eugeneswalker**
Unify when possible. Interested to see how this goes. Will everything continue to build when the environment is unified?

FYI @wspear @alalazo @becker33

### CODE REVIEW

**Comment by eugeneswalker** [comment]
Just added `ecp-data-vis-sdk` FYI @kwryankrattiger

File: share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml [hunk 1]
```diff
@@ -149,15 +143,11 @@ spack:
...
-  - umpire
   - upcxx
-  - veloc
-  - vtk-m
   - wannier90
-  - zfp
 
   # CUDA
-  - adios2 +cuda cuda_arch=80
+  - ecp-data-vis-sdk +adios2 +ascent +cinema +cuda +darshan +faodel +hdf5 +paraview +pnetcdf ~rocm ~sensei +sz +unifyfs +veloc +visit +vtkm +zfp cuda_arch=80
```
Comment (kwryankrattiger) [inline share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml]: Paraview needs to be disabled here, there are known issues when building with cuda.

Comment (eugeneswalker) [reply share/spack/gitlab/cloud_pipelines/stacks/e4s/spack.yaml]: done!

**Comment by eugeneswalker** [comment]
It looks like the environment @ https://github.com/spack/spack/pull/31940/commits/f9096f497fee1d0ad2e8b8955fbae85922ba8d0e is concretizing to include the `opengl` placeholder package, which doesn't work in the container:
* See this job for reference: https://gitlab.spack.io/spack/spack/-/jobs/2947345

I will add `^mesa +osmesa ~glx`  to the `ecp-data-vis-sdk` specs and hope that `unify: when_possible` causes it to replace `opengl` placeholder everywhere.

**Comment by eugeneswalker** [comment]
Quite a few failures:

* [(specs) gotcha/p6zl33b 1.0.3 gcc@11.2.0 linux-ubuntu22.04-x86_64 New PR testing workflow](https://gitlab.spack.io/spack/spack/-/jobs/2955784)
```
...
     129    /builds/spack/spack/lib/spack/env/gcc/gcc -O2 -g -DNDEBUG -rdynamic
             CMakeFiles/symb_look.dir/symbolLookup.c.o -o symb_look  -Wl,-rpath
            ,/tmp/root/spack-stage/spack-stage-gotcha-1.0.3-p6zl33bw5amx6wkwl4k
            hg5iu6xwbbbrp/spack-build-p6zl33b/src/example/minimal:/tmp/root/spa
            ck-stage/spack-stage-gotcha-1.0.3-p6zl33bw5amx6wkwl4khg5iu6xwbbbrp/
            spack-build-p6zl33b/src libwrap_me.so ../../libgotcha.so.2.0.2
  >> 130    /usr/bin/ld: ../../libgotcha.so.2.0.2: undefined reference to `_dl_
            sym'
  >> 131    collect2: error: ld returned 1 exit status
  >> 132    make[2]: *** [src/example/minimal/CMakeFiles/symb_look.dir/build.ma
            ke:102: src/example/minimal/symb_look] Error 1
...
```

* [(specs) plasma/fbgu447 21.8.29 gcc@11.2.0 linux-ubuntu22.04-x86_64 New PR testing workflow](https://gitlab.spack.io/spack/spack/-/jobs/2955835)
```
...
1 error found in build log:
     17    -- Could NOT find MKL (missing: MKL_INCLUDE_DIRS MKL_LIBRARIES)
     18    -- Could NOT find OpenBLAS (missing: OpenBLAS_INCLUDE_DIRS OpenBLAS_
           LIBRARIES)
     19    -- Could NOT find NetlibCblas (missing: NetlibCblas_INCLUDE_DIRS Net
           libCblas_LIBRARIES)
     20    -- Looking for cblas_cgemm
     21    -- Looking for cblas_cgemm - not found
     22    -- Could NOT find Accelerate (missing: Accelerate_INCLUDE_DIRS Accel
           erate_LIBRARIES)
...
     24      Could NOT find CBLAS (missing: CBLAS_INCLUDE_DIRS CBLAS_LIBRARIES)
...
     27      Modules/FindCBLAS.cmake:69 (find_package_handle_standard_args)
     28      CMakeLists.txt:74 (find_package)
     29
...
```

* [(specs) llvm/2usjisl 14.0.6 gcc@11.2.0 linux-ubuntu22.04-x86_64 New PR testing workflow](https://gitlab.spack.io/spack/spack/-/jobs/2955977) - job log truncated

* [(specs) phist/nqv75gw 1.9.5 gcc@11.2.0 linux-ubuntu22.04-x86_64 New PR testing workflow](https://gitlab.spack.io/spack/spack/-/jobs/2955837)
```
==> [2022-08-06-16:26:53.383187] phist: Executing phase: 'cmake'
==> [2022-08-06-16:26:53.492425] Error: NoHeadersError: Unable to locate nvhpc headers in /home/software/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spa/linux-ubuntu22.04-x86_64/gcc-11.2.0/nvhpc-22.7-zoqm7b6df3gao7k23voslunitvvibrjh

/builds/spack/spack/var/spack/repos/builtin/packages/phist/package.py:235, in cmake_args:
        232        lapacke_libs = (
        233            spec["lapack:c"].libs + spec["blas:c"].libs + find_system_libraries(["libm"])
        234        ).joined(";")
  >>    235        lapacke_include_dir = spec["lapack:c"].headers.directories[0]
        236
        237        args = [
        238            "-DPHIST_USE_CCACHE=OFF",
```

* [(specs) butterflypack/2oeiboa 2.1.1 gcc@11.2.0 linux-ubuntu22.04-x86_64 New PR testing workflow](https://gitlab.spack.io/spack/spack/-/jobs/2955901)
```
...
  >> 2058    /usr/bin/ld: CMakeFiles/ie3d.dir/EMSURF_Driver.f90.o: undefined re
             ference to symbol 'omp_get_wtime_@@@@VERSION'
  >> 2059    /usr/bin/ld: /home/software/spack/__spack_path_placeholder__/__spa
             ck_path_placeholder__/__spack_path_placeholder__/__spack_path_plac
             eholder__/__spack_path_placeholder__/__spack_path_placeholder__/__
             spack_path_placeholder__/__spack_path_placeholder__/__spack_path_p
             laceholder__/__spack_path_placeholder__/__spack_path_placeholder__
             /__spack_path_placeholder__/__spack_path_placeholder__/__spack_pat
             h_placeholder__/__spack_path_placeholder__/__spack_path_placeholde
             r__/__spack_path_placeholder__/__spack_path_placeholder__/__spa/li
             nux-ubuntu22.04-x86_64/gcc-11.2.0/nvhpc-22.7-zoqm7b6df3gao7k23vosl
             unitvvibrjh/Linux_x86_64/22.7/compilers/lib/libnvomp.so: error add
             ing symbols: DSO missing from command line
  >> 2060    collect2: error: ld returned 1 exit status
  >> 2061    make[2]: *** [EXAMPLE/CMakeFiles/ie3d.dir/build.make:124: EXAMPLE/
             ie3d] Error 1
...
```

* [(specs) butterflypack/lwyogu2 2.1.1 gcc@11.2.0 linux-ubuntu22.04-x86_64 New PR testing workflow](https://gitlab.spack.io/spack/spack/-/jobs/2956283)
```
...
  >> 2059    /usr/bin/ld: CMakeFiles/ie3d.dir/EMSURF_Driver.f90.o: undefined re
             ference to symbol 'omp_get_wtime_@@@@VERSION'
  >> 2060    /usr/bin/ld: /home/software/spack/__spack_path_placeholder__/__spa
             ck_path_placeholder__/__spack_path_placeholder__/__spack_path_plac
             eholder__/__spack_path_placeholder__/__spack_path_placeholder__/__
             spack_path_placeholder__/__spack_path_placeholder__/__spack_path_p
             laceholder__/__spack_path_placeholder__/__spack_path_placeholder__
             /__spack_path_placeholder__/__spack_path_placeholder__/__spack_pat
             h_placeholder__/__spack_path_placeholder__/__spack_path_placeholde
             r__/__spack_path_placeholder__/__spack_path_placeholder__/__spa/li
             nux-ubuntu22.04-x86_64/gcc-11.2.0/nvhpc-22.7-zoqm7b6df3gao7k23vosl
             unitvvibrjh/Linux_x86_64/22.7/compilers/lib/libnvomp.so: error add
             ing symbols: DSO missing from command line
  >> 2061    collect2: error: ld returned 1 exit status
  >> 2062    make[2]: *** [EXAMPLE/CMakeFiles/ie3d.dir/build.make:124: EXAMPLE/
             ie3d] Error 1
...
```

* [(specs) py-numpy/k4qrira 1.21.5 gcc@11.2.0 linux-ubuntu22.04-x86_64 New PR testing workflow](https://gitlab.spack.io/spack/spack/-/jobs/2955836)
```
...
==> [2022-08-06-18:41:18.206965] py-numpy: Executing phase: 'install'
==> [2022-08-06-18:41:18.265980] Error: NoHeadersError: Unable to locate nvhpc headers in /home/software/spack/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spack_path_placeholder__/__spa/linux-ubuntu22.04-x86_64/gcc-11.2.0/nvhpc-22.7-zoqm7b6df3gao7k23voslunitvvibrjh

/builds/spack/spack/var/spack/repos/builtin/packages/py-numpy/package.py:222, in set_blas_lapack:
        219        blas_headers = HeaderList([])
        220        if "+blas" in spec:
        221            blas_libs = spec["blas"].libs
  >>    222            blas_headers = spec["blas"].headers
        223
        224        lapack_libs = LibraryList([])
        225        lapack_headers = HeaderList([])
...
```

* [(specs) slate/l6xobus 2022.06.00 gcc@11.2.0 linux-ubuntu22.04-x86_64 New PR testing workflow](https://gitlab.spack.io/spack/spack/-/jobs/2955838)
```
...
     266     In file included from /tmp/root/spack-stage/spack-stage-slate-2022
             .06.00-l6xobus4d53nui5imhde3i4znvm4m7p7/spack-src/src/gbtrs.cc:11:
  >> 267     /tmp/root/spack-stage/spack-stage-slate-2022.06.00-l6xobus4d53nui5
             imhde3i4znvm4m7p7/spack-src/src/internal/internal.hh:17:14: fatal 
             error: cblas.h: No such file or directory
     268        17 |     #include <cblas.h>
     269           |              ^~~~~~~~~
     270     compilation terminated.
  >> 271     make[2]: *** [CMakeFiles/slate.dir/build.make:259: CMakeFiles/slat
             e.dir/src/gbtrs.cc.o] Error 1
...
```

* [(specs) vtk-h/467hiqi 0.8.1 gcc@11.2.0 linux-ubuntu22.04-x86_64 New PR testing workflow](https://gitlab.spack.io/spack/spack/-/jobs/2955820)
```
...
-- The C compiler identification is GNU 11.2.0
-- The CXX compiler identification is GNU 11.2.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/gcc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/g++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
CMake Error at CMakeLists.txt:66 (message):
  Static libraries are required when building with CUDA


-- Configuring incomplete, errors occurred!
...
```

**Comment by eugeneswalker** [comment]
I think trying to switch to `unify: when_possible` may be a bit too much to do at the same time as we integrate the `ecp-data-vis-sdk` metapackage. I created another PR that *only* makes the switch to `ecp-data-vis-sdk` and leaves the unification strategy alone:
* https://github.com/spack/spack/pull/31974

**Pull request closed without merging**